test/test_dynamo_cudagraphs.py

```diff
@@ -0,0 +1,146 @@
+import torch
+from torch.utils._pytree import tree_map
+from torch.fx.passes.backends.cudagraphs import partition_cudagraphs
+from torch.testing._internal.common_utils import (
+    TestCase,
+    run_tests,
+)
+try:
+    import torchdynamo
+    TEST_DYNAMO = True
+except ImportError:
+    TEST_DYNAMO = False
+
+TEST_CUDA = torch.cuda.is_available()
+
+if not TEST_CUDA or not TEST_DYNAMO:
+    print('CUDA or dynamo not available, skipping tests', file=sys.stderr)
+    TestCase = object  # noqa: F811
+
+torchdynamo.config.verify_correctness = True
+
+class RunCudaGraph():
+    warmed_up = False
+
+    # these are all None or all filled
+    graph = None
+    static_inputs = None
+    static_outputs = None
+
+    def __call__(self, gm, *args):
+        def cloner(t):
+            if isinstance(t, torch.Tensor):
+                return t.clone()
+            else:
+                return t
+
+        # TODO: once we've recorded here, we'd like to replace this with
+        # compiled bytecode that copies into static, replays the cuda graph,
+        # then copies out.  First condition is the hotpath, needs optimizing
+        if self.graph is not None:
+            assert len(args) == len(self.static_inputs)
+            for dst, src in zip(self.static_inputs, args):
+                dst.copy_(src)
+            self.graph.replay()
+            return tree_map(cloner, self.static_outputs)
+
+        elif self.warmed_up:
+            # record
+            self.static_inputs = [x.clone() for x in args]
+            self.graph = torch.cuda.CUDAGraph()
+            with torch.cuda.graph(self.graph):
+                # TODO: this is wrong but required due to _wrapped_call
+                self.static_outputs = super(type(gm), gm).__call__(*self.static_inputs)
+            return tree_map(cloner, self.static_outputs)
+
+        else:
+            # warmup
```

Comment:
what does warmup actually do for graph recording?

our docs suggest to warm up for "a few iterations" (in their example they use 3) but idk what for.   
https://pytorch.org/blog/accelerating-pytorch-with-cuda-graphs/#api-example

Reply:
I think it's ddp, not nccl, but I'm not sure. Cudnn benchmarking throws an error if someone tries to capture it.

Reply:
oops you're right